Sources/PackageModel/DependencyMapper.swift

```diff
@@ -29,143 +28,220 @@ public struct DefaultDependencyMapper: DependencyMapper {
     }
 
-    public static func sanitizeDependencyLocation(fileSystem: FileSystem, packageKind: PackageReference.Kind, dependencyLocation: String) throws -> String {
-        if dependencyLocation.hasPrefix("~/") {
-            // If the dependency URL starts with '~/', try to expand it.
-            return try AbsolutePath(validating: String(dependencyLocation.dropFirst(2)), relativeTo: fileSystem.homeDirectory).pathString
-        } else if dependencyLocation.hasPrefix(filePrefix) {
-            // FIXME: SwiftPM can't handle file locations with file:// scheme so we need to
-            // strip that. We need to design a Location data structure for SwiftPM.
-            let location = String(dependencyLocation.dropFirst(filePrefix.count))
-            let hostnameComponent = location.prefix(while: { $0 != "/" })
-            guard hostnameComponent.isEmpty else {
-              if hostnameComponent == ".." {
-                  throw DependencyMappingError.invalidFileURL("file:// URLs cannot be relative, did you mean to use '.package(path:)'?")
-              }
-              throw DependencyMappingError.invalidFileURL("file:// URLs with hostnames are not supported, are you missing a '/'?")
+    fileprivate var nameForTargetDependencyResolutionOnly: String? {
+        switch self.kind {
+        case .fileSystem(let name, _):
+            return name
+        case .sourceControl(let name, _, _):
+            return name
+        case .registry:
+            return .none
+        }
+    }
+
+    fileprivate var sourceControlRequirement: PackageDependency.SourceControl.Requirement {
+        get throws {
+            switch self.kind {
+            case .fileSystem:
+                throw DependencyMappingError.invalidMapping("invalid mapping of file system to source control, requirement information mismatch")
+            case .sourceControl(_, _, let requirement):
+                return requirement
+            case .registry(_, let requirement):
+                return .init(requirement)
             }
-            return try AbsolutePath(validating: location).pathString
-        } else if parseScheme(dependencyLocation) == nil {
-            // If the URL has no scheme, we treat it as a path (either absolute or relative to the base URL).
-            switch packageKind {
-            case .root(let packagePath), .fileSystem(let packagePath), .localSourceControl(let packagePath):
-                return try AbsolutePath(validating: dependencyLocation, relativeTo: packagePath).pathString
-            case .remoteSourceControl, .registry:
-                // nothing to "fix"
-                return dependencyLocation
+        }
+    }
+
+    fileprivate var registryRequirement: PackageDependency.Registry.Requirement {
+        get throws {
+            switch self.kind {
+            case .fileSystem:
+                throw DependencyMappingError.invalidMapping("invalid mapping of file system to registry, requirement information mismatch")
```

Comment:
Ditto: can we provide more details on the mismatch in the error message to make it actionable?